Feature/grading instructions/display sgi in assessment view fix

HanyaElhashemy opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~~Server: I added multiple integration tests (Spring) related to the features~~
- [x] ~~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~~
- [x] ~~Server: I implemented the changes with a good performance and prevented too many database calls~~
- [x] ~~Server: I documented the Java code using JavaDoc style.~~
- [ ] Client: I added multiple integration tests (Jest) related to the features
- [x] ~~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~~
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
As multiple tutors asses student's submission, the assessment becomes inconsistent.
Tutors used to insert free text feedback and apply their individual interpretation of the grading instructions. This leads to the feedback being sometimes faulty or incomplete. Therefore, instructors can predefine feedback text that tutors can directly drag and drop. This also saves tutors time as they don't need to type the same feedback over and over again.
### Description
<!-- Describe your changes in detail -->
The structured grading instructions are displayed at three different Places:
a.) At the exercise dashboard before Tutors participate in the assessment
b.) At the side panel of the exercise dashboard after tutors submit that they read the instructions
c.) beside students submission in the tutor assessment view
The instructions could be dragged and dropped to the feedback element.
The instructions div color depends on the score: 
< 0 : red
=0 : yellow
>0 : green

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
make sure that you are signed in as an admin:
1. Create a file-upload or text exercise with structured grading instructions
2. Create a student's submission for the exercise you created
3. Navigate to the tutor dashboard -> the structured grading instructions should be displayed there
4. Start assessing the submission
5. Drag and drop an instruction to the feedback area -> the score and feedback text should be applied
6. Drag and drop an instruction to the general feedback field as well


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
a.)
![image](https://user-images.githubusercontent.com/50635005/77358960-25a88f80-6d4b-11ea-8872-6143a5ff3341.png)
b.)
![image](https://user-images.githubusercontent.com/50635005/77359009-3d801380-6d4b-11ea-973a-e24ad15e9e54.png)
c.)
![image](https://user-images.githubusercontent.com/50635005/77359206-833cdc00-6d4b-11ea-9b09-30dd34ac671e.png)
Drag & Drop to feedback element
![image](https://user-images.githubusercontent.com/50635005/77359249-96e84280-6d4b-11ea-8dd1-afaf63dc33a4.png)
![image](https://user-images.githubusercontent.com/50635005/77359284-a4053180-6d4b-11ea-8473-b485e3fc3bb5.png)
Drag & Drop to general feedback field
![image](https://user-images.githubusercontent.com/50635005/77359353-ba12f200-6d4b-11ea-9797-0d6a0cef3655.png)
![image](https://user-images.githubusercontent.com/50635005/77359380-c5feb400-6d4b-11ea-9497-3e7234e5ec60.png)

If an element is missing in the text exercise submission tutors can drop the instruction to the general feedback field but in a separate PR this should be addressed by enabling tutors to create a proper feedback element.

sascha11110 commented on src/main/webapp/i18n/de/exercise.json:
```diff
@@ -33,6 +33,7 @@
             "sampleSolution": "Beispiel Lösung",
             "gradingInstructions": "Benotungshinweise",
             "structuredGradingInstructions": "Strukturierte Benotungshinweise",
+            "DragDropInstruction": "Drag & Drop Bewetungshinweis zum Feedback Element",
```
Spelling mistake :-)

sascha11110 commented on src/main/webapp/app/assessment/assessment-detail/assessment-detail.component.ts:
```diff
@@ -42,4 +42,17 @@ export class AssessmentDetailComponent {
+    allowDrop(ev: any) {
+        ev.preventDefault();
+    }
+    drop(ev: any) {
```
You might want to refactor the duplicated code for `allowDrop()` and `drop()` **(1)**.

HanyaElhashemy replied:
This `drop()` is for the feedback elements with reference. The other methods are for general feedback. I refactored the general feedback drop method further and used the assessment-general-feedback component for the text and file-upload exercises as well. Thank you for noticing this.

sascha11110 commented on src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:
```diff
@@ -489,4 +489,18 @@ export class FileUploadAssessmentComponent implements OnInit, AfterViewInit, OnD
+    allowDrop(ev: any) {
+        ev.preventDefault();
+    }
+    drop(ev: any) {
```
You might want to refactor the duplicated code for `allowDrop()` and `drop()` **(2)**.

sascha11110 commented on src/main/webapp/app/exercises/text/assess/text-assessment.component.ts:
```diff
@@ -504,4 +504,19 @@ export class TextAssessmentComponent implements OnInit, OnDestroy, AfterViewInit
+    allowDrop(ev: any) {
+        ev.preventDefault();
+    }
+    drop(ev: any) {
```
You might want to refactor the duplicated code for `allowDrop()` and `drop()` **(3)**.

HanyaElhashemy commented:
> Hey Hanya, overall good job!

Thank you for your review :)

> Unfortunately the instructions are not shown in the tutor assessment view (they appear in the other two places):

Could you please explain the steps that lead to this in order for me to recreate this issue? Is there a console error?

> <img alt="missing" width="632" src="https://user-images.githubusercontent.com/15168372/77459497-19cad500-6e00-11ea-9aa2-aa6d363e1c6c.png">
> 
> In addition, it would be good if you would add some margin in the places marked in red. You could make the blue marked column wider, then whole space would be better used :-)
> 
> ![margin](https://user-images.githubusercontent.com/15168372/77459867-9d84c180-6e00-11ea-8f79-705136ab04a8.jpg)

The example text is very long which gives the impression that the blue marked column needs more place, however, in reality, this column is used for the grading scale e.g. "good" or "average" or "poor" so the cell content is only one word or a percentage, therefore it does not need extra space imo.

I removed the margins after our discussion when you mentioned that the structured grading instructions are taking a lot of space

alextsakpinis commented:
Hello Hanya,
great work at first. Could you change the naming of that element. Could be better

![image](https://user-images.githubusercontent.com/23211992/77511716-f80a3600-6e71-11ea-9c05-5efebf9cb52c.png)

alextsakpinis commented:
Same for me. Somehow the grading instructions are not shown. I followed the steps in your description. The only thing i also did was to add more instrucions once I had already started an assessment. Don't know if that matters?

![image](https://user-images.githubusercontent.com/23211992/77511925-4d464780-6e72-11ea-8d10-38bf3d6df2d6.png)

Here you  can see that I inserted a few..
![image](https://user-images.githubusercontent.com/23211992/77511985-6818bc00-6e72-11ea-879a-c5083dfab061.png)

HanyaElhashemy commented on src/main/webapp/app/assessment/assessment-general-feedback/assessment-general-feedback.component.ts:
```diff
@@ -22,4 +22,16 @@ export class AssessmentGeneralFeedbackComponent {
+    allowDrop(ev: any) {
+        ev.preventDefault();
+    }
+    drop(ev: any) {
```
drop and allowDrop can't be merged as they are triggered on different events. If you have a look at the html you will see that allowDrop is triggered on dragover and not on drop.

HanyaElhashemy commented:
> Hello Hanya,
> great work at first. Could you change the naming of that element. Could be better
> 
> ![image](https://user-images.githubusercontent.com/23211992/77511716-f80a3600-6e71-11ea-9c05-5efebf9cb52c.png)

Hi Alex,
Good catch, this is a jhiTranslate mistake on develop but I fixed it in my branch!

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -216,6 +223,8 @@ else if (!authCheckService.isAtLeastInstructorForExercise(exercise)) {
             @RequestParam(value = "lock", defaultValue = "false") boolean lockSubmission) {
         log.debug("REST request to get a file upload submission without assessment");
         final Exercise fileUploadExercise = exerciseService.findOneWithAdditionalElements(exerciseId);
+        List<GradingCriterion> gradingCriteria = gradingCriterionService.findByExerciseIdWithEagerGradingCriteria(exerciseId);
```
can you load this together with the exercise above?

krusche commented on src/main/webapp/app/assessment/assessment-detail/assessment-detail.component.ts:
```diff
@@ -42,4 +42,17 @@ export class AssessmentDetailComponent {
     public get sanitizedText(): string {
         return sanitize(this.text);
     }
+    allowDrop(ev: any) {
```
can you write event. Do we have a reasonable type?

krusche commented on src/main/webapp/app/assessment/assessment-detail/assessment-detail.component.ts:
```diff
@@ -42,4 +42,17 @@ export class AssessmentDetailComponent {
+    allowDrop(ev: any) {
+        ev.preventDefault();
+    }
+    drop(ev: any) {
```
can you write event. Do we have a reasonable type?

krusche commented on src/main/webapp/app/assessment/assessment-general-feedback/assessment-general-feedback.component.ts:
```diff
@@ -22,4 +22,16 @@ export class AssessmentGeneralFeedbackComponent {
         this.feedbackClone.detailText = trimmedText.length > 0 ? trimmedText : null;
         this.feedbackChange.emit(this.feedbackClone);
     }
+    allowDrop(ev: any) {
```
can you write event. Do we have a reasonable type?

HanyaElhashemy replied:
Yes, DragEvent so I changed it accordingly, thanks for pointing this out.

krusche commented on src/main/webapp/app/assessment/assessment-general-feedback/assessment-general-feedback.component.ts:
```diff
@@ -22,4 +22,16 @@ export class AssessmentGeneralFeedbackComponent {
+    allowDrop(ev: any) {
+        ev.preventDefault();
+    }
+    drop(ev: any) {
```
can you write event. Do we have a reasonable type?

HanyaElhashemy commented:
@krusche I applied your comments, would you be so kind and have another look. Thanks

madwau commented:
I just tested it on the test server and it works nicely, well done!

A couple of smaller things I noticed:
1. When enabling SGIs in the exercise create form, I wasn't really sure what the difference is between a grading criterion and a grading instruction. An easy way to make this a little more clear I think would be to include a grading criterion in the example SGIs that get auto-generated inititially.
2. When going back to edit a previously created exercise with SGIs, the SGI switch is not flipped and it looks like the information is lost on first sight. Only when switching into SGI mode, the instructions come back. Simply put: the state of the switch needs to be initialized correctly.
3. When changing the text of the SGI textarea, a console log happens. This is not part of this PR but maybe it can be removed here anyway.
4. I think most of the attributes of a grading instruction are self-explanatory, however, I wasn't really sure what `count` does. I was assuming that this is the max number of times this instruction can be applied on one submission but there seems to be no such restriction when assessing. Can you clarify?
5. When dragging an item into the general feedback area, it shows the score in the textarea but it is not actually applied. This is probably intended but I can see how some people might expect this score to be applied as well.

HanyaElhashemy commented:
> I just tested it on the test server and it works nicely, well done!
> 
Thank you so much for your review!

> A couple of smaller things I noticed:
> 
> 1. When enabling SGIs in the exercise create form, I wasn't really sure what the difference is between a grading criterion and a grading instruction. An easy way to make this a little more clear I think would be to include a grading criterion in the example SGIs that get auto-generated inititially.

There is an example provided when you first open the markdown editor. However, this example shows a separate Instruction. If I provided an Instruction that belongs to a criterion instead then the user will not know that the option of separate instructions exists.  What I could do is provide a hint, or add a second example that displays a criterion but in this case, the user will be initially faced by a wall of text which is not that nice in terms of usability. What do the others think?

> 2. When going back to edit a previously created exercise with SGIs, the SGI switch is not flipped and it looks like the information is lost on first sight. Only when switching into SGI mode, the instructions come back. Simply put: the state of the switch needs to be initialized correctly.

You mean I should add a hint that the initialization of the switch is the grading instructions, not the SGI?

> 3. When changing the text of the SGI textarea, a console log happens. This is not part of this PR but maybe it can be removed here anyway.

Good catch! I removed it.

> 4. I think most of the attributes of a grading instruction are self-explanatory, however, I wasn't really sure what `count` does. I was assuming that this is the max number of times this instruction can be applied on one submission but there seems to be no such restriction when assessing. Can you clarify?

Good point! I renamed it to "max count in score". This parameter as you said specifies how often points of the instruction should be added to the final score. The instruction could be used more times than the nr. specified but in this case, the points will not be added

> 5. When dragging an item into the general feedback area, it shows the score in the textarea but it is not actually applied. This is probably intended but I can see how some people might expect this score to be applied as well.

I removed the drag and drop to the general feedback and will be working on a different way to add feedback for a missing submission element in a separate PR.

HanyaElhashemy commented:
> I tested your changes and everything works as expected :-) Unfortunately I found a _major_ usability problem.
> 
> When the assessing tutor has a smaller screen it is not possible to drag the feedback of the grading instruction to the box where it should be added. The problem is that one cannot scroll while _holding_ the grading instruction and therefore the box is not reachable if it is outside of the current viewport.
> 
> The problem occurs in Chrome, Firefox and Safari on Mac.

Thank you for your review. I was not able to recreate your issue, please have a look at my screencast where I simulated a small screen on chrome and firefox.
![Animated GIF-downsized_large (1)](https://user-images.githubusercontent.com/50635005/78889936-f8efba00-7a64-11ea-871d-5d1edcdc2605.gif)
Also, note that there is a new text exercise version where the feedback will be inline so there will be no reason to scroll down.

madwau commented:
> There is an example provided when you first open the markdown editor. However, this example shows a separate Instruction. If I provided an Instruction that belongs to a criterion instead then the user will not know that the option of separate instructions exists. What I could do is provide a hint, or add a second example that displays a criterion but in this case, the user will be initially faced by a wall of text which is not that nice in terms of usability. What do the others think?

I see your point. I think I'd tend to favor a longer example that showcases all the different usages since understanding how the feature works is probably the single most important requirement ahead of aesthetics and so on but yes, would be interesting to hear the perspective of others on this.

> You mean I should add a hint that the initialization of the switch is the grading instructions, not the SGI?

No, I mean the state of the switch is not correct when loading the page. If I create an exercise with SGI mode and then go back to it later, the switch won't be flipped initially, making it look like the SGIs are gone.

> I removed the drag and drop to the general feedback and will be working on a different way to add feedback for a missing submission element in a separate PR.

Sounds good :)

HanyaElhashemy commented:
> > There is an example provided when you first open the markdown editor. However, this example shows a separate Instruction. If I provided an Instruction that belongs to a criterion instead then the user will not know that the option of separate instructions exists. What I could do is provide a hint, or add a second example that displays a criterion but in this case, the user will be initially faced by a wall of text which is not that nice in terms of usability. What do the others think?
> 
> I see your point. I think I'd tend to favor a longer example that showcases all the different usages since understanding how the feature works is probably the single most important requirement ahead of aesthetics and so on but yes, would be interesting to hear the perspective of others on this.
> 
I implemented your suggestion since no one suggested a different approach. It was not much text after all and I think it showcases the functionality now better. Thanks for the suggestion!

> > You mean I should add a hint that the initialization of the switch is the grading instructions, not the SGI?
> 
> No, I mean the state of the switch is not correct when loading the page. If I create an exercise with SGI mode and then go back to it later, the switch won't be flipped initially, making it look like the SGIs are gone.
> 
It is not an either or relationship. You could have both Grading Instructions (GI) and SGI. So the question here is if we have both GI and SGI which one should we display on init in the exercise edit page and if we have none (for example for new exercises) ? This is why I intentionally set GI as default. The other approach would be to save the state of the toggle in the database or local storage in order to set this state on init **or** display both editors and remove the toggle (which I prefer to keep it simple) 
@krusche what do you think?
> > I removed the drag and drop to the general feedback and will be working on a different way to add feedback for a missing submission element in a separate PR.
> 
> Sounds good :)

krusche commented on src/main/webapp/app/assessment/assessment-instructions/assessment-instructions/assessment-instructions.component.ts:
```diff
@@ -55,4 +59,7 @@ export class AssessmentInstructionsComponent {
             this.sampleSolution = this.markdownService.safeHtmlForMarkdown(sampleSolutionMarkdown);
         }
     }
+    drag(ev: any, instruction: GradingInstruction) {
```
please add a JSDoc comment here, that this method connects the GradingInstruction with the Feedback

krusche commented on src/main/webapp/app/assessment/assessment-detail/assessment-detail.component.ts:
```diff
@@ -42,4 +42,17 @@ export class AssessmentDetailComponent {
+    allowDrop(event: DragEvent) {
+        event.preventDefault();
+    }
+    drop(event: any) {
```
please add a JSDoc comment here, that this method connects the GradingInstruction with the Feedback
link/reference the corresponding drag method in the JSDoc

TobiasPr commented:
Wanted to warm up the discussion on this. I also had the problem that I find the toggle missleading, because it indicates an either or relationship in my opinion ( I know it mostly from settings where one can decide to enable or disable something). Maybe a tabview is more understandable since it is not a either or realationship
![image](https://user-images.githubusercontent.com/22955066/81050548-7372f480-8ec0-11ea-80bf-17b5b0434df1.png)

madwau commented:
@TobiasPr True, tabs would be the perfect semantic UI element for this scenario.
